### ISSUE TITLE
Jnp.concatenate does work right

### ISSUE THREAD

**Issue by ruili3** [google/jax#14490]
### Description

Hi, I'm writing a deep neural network training code with Jax, Flax, and Optax. I observed a wired thing when I use jnp. concatenate() to aggregate two matrices into a certain dimension. The operation does not yield the right results as seen in jax.debug.print().

I conducted the operation inside the loss function, which is contained in another training function decorated with jax.pmap(). Part of the code is shown below:
```python
def compute_losses_non_parallel(video_init_fr, video_init_adj_fr, warped_video_fr, disp_pred_fr):
  # video_init_fr- shape:n,fr,h,w,3
  # video_init_adj_fr - shape:n,fr, 2,h,w,3
  # warpped_video_fr - shape:n,fr, adj-2,disp-4,h,w,3
  # disp_pred_fr - shape: n,fr, h,w,4
  loss_total=0
  losses = {}
  total_scales = 4
  fr_num = video_init_fr.shape[1]


  for ii in range(fr_num):

    video_init = video_init_fr[:,ii]
    video_init_adj = video_init_adj_fr[:,ii]
    warped_video = warped_video_fr[:,ii]
    disp_pred = disp_pred_fr[:,ii]

    for tt in range(total_scales):
      loss = 0
      reproj_losses=[]
      for kk in range(2):
         # b,h,w,1
        reproj_losses.append(compute_photometric_loss(warped_video[:,kk,tt],video_init))

        if tt==3 and ii==1:
          jax.debug.print("===scale-4-fr1-adj:{b} loss:{a}", a=compute_photometric_loss(warped_video[:,kk,tt],video_init), b=kk)
        
      
      # b,h,w,2
      reproj_losses = jnp.concatenate(reproj_losses, axis=3)

      if tt==3 and ii==1:
        jax.debug.print("===reproj_losses:{a}", a=reproj_losses)
```
During running, part of the `scale-4-fr1-adj:0` loss and `scale-4-fr1-adj:1` loss are:
```python
===scale-4-fr1-adj:0 loss:[[[[0.8525491 ]
   [0.00745098]
   [0.01235293]
   ...
   [0.01306295]
   [0.00316211]
   [0.0361502 ]]

  [[0.02039214]
   [0.01627451]
   [0.01862744]
   ...
   [0.01470877]
   [0.01998307]
   [0.01764817]]
```

```python
===scale-4-fr1-adj:1 loss:[[[[0.4105921 ]
   [0.04968597]
   [0.0486871 ]
   ...
   [0.02002719]
   [0.04886725]
   [0.01558158]]

  [[0.01098038]
   [0.5680393 ]
   [0.06133314]
   ...
   [0.0539965 ]
   [0.0697121 ]
   [0.02950328]]
```
When conducting `reproj_losses = jnp.concatenate(reproj_losses, axis=3)`, we got
```python
===reproj_losses:[[[[0.8525491  0.00745098]
   [0.01235293 0.57176477]
   [0.63534826 0.23562063]
   ...
   [0.01507517 0.01095143]
   [0.02490095 0.01470877]
   [0.01998307 0.01764817]]

  [[0.0262745  0.0298039 ]
   [0.02882352 0.02431372]
   [0.01588232 0.01411766]
```
It is not a normal operation from jnp.concatenate(), but more like a reshape() operation of the loss maps. The output shapes are shown in the source code. Can you help to provide some hints on what caused this? Thank you very much!

### What jax/jaxlib version are you using?

jax 0.3.24; jaxlib 0.3.24+cuda11.cudnn82; flax0.4.2; optax 0.1.4

### Which accelerator(s) are you using?

GPU

### Additional system info

Linux, python 3.8.5

### NVIDIA GPU info

```
Wed Feb 15 18:44:09 2023       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 515.86.01    Driver Version: 515.86.01    CUDA Version: 11.7     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Quadro RTX 6000     On   | 00000000:21:00.0 Off |                  Off |
| 33%   29C    P8    11W / 260W |      1MiB / 24576MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   1  Quadro RTX 6000     On   | 00000000:A1:00.0 Off |                  Off |
| 33%   30C    P8    10W / 260W |      1MiB / 24576MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+
```

**Comment by jakevdp** [comment]
Hi - thanks for the report. I'm having trouble understanding what your expected output is. Can you provide a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) showing the output of `jnp.concatenate` and how it differs from what you expect?

**Comment by jakevdp** [comment]
Closing for lack of information. Please let us know if you're still having this issue!
